tools/src/main/java/org/apache/kafka/tools/consumer/ConsoleConsumer.java

```diff
@@ -0,0 +1,233 @@
+    }
+
+    static void addShutdownHook(ConsumerWrapper consumer, ConsoleConsumerOptions conf) {
+        Exit.addShutdownHook("consumer-shutdown-hook", () -> {
+            try {
+                consumer.wakeup();
+                SHUTDOWN_LATCH.await();
+            } catch (Throwable t) {
+                LOG.error("Exception while running shutdown hook " + t.getMessage());
+            }
+            if (conf.enableSystestEventsLogging()) {
+                System.out.println("shutdown_complete");
+            }
+        });
+    }
+
+    static void process(int maxMessages, MessageFormatter formatter, ConsumerWrapper consumer, PrintStream output, boolean skipMessageOnError) {
+        while (messageCount < maxMessages || maxMessages == -1) {
+            ConsumerRecord<byte[], byte[]> msg;
+            try {
+                msg = consumer.receive();
+            } catch (WakeupException we) {
+                LOG.trace("Caught WakeupException because consumer is shutdown, ignore and terminate.");
+                // Consumer will be closed
+                return;
+            } catch (Throwable t) {
+                LOG.error("Error processing message, terminating consumer process: ", t);
+                // Consumer will be closed
+                return;
+            }
+            messageCount += 1;
+            try {
+                formatter.writeTo(new ConsumerRecord<>(msg.topic(), msg.partition(), msg.offset(), msg.timestamp(), msg.timestampType(),
+                    0, 0, msg.key(), msg.value(), msg.headers(), Optional.empty()), output);
+            } catch (Throwable t) {
+                if (skipMessageOnError) {
+                    LOG.error("Error processing message, skipping this message: ", t);
+                } else {
+                    // Consumer will be closed
+                    throw t;
+                }
+            }
+            if (checkErr(output)) {
+                // Consumer will be closed
+                return;
+            }
+        }
+    }
+
+    static void reportRecordCount() {
+        System.err.println("Processed a total of " + messageCount + " messages");
+    }
+
+    static boolean checkErr(PrintStream output) {
+        boolean gotError = output.checkError();
+        if (gotError) {
+            // This means no one is listening to our output stream anymore, time to shut down
+            System.err.println("Unable to write to standard out, closing consumer.");
+        }
+        return gotError;
+    }
+
+    public static class ConsumerWrapper {
+        final Optional<String> topic;
+        final OptionalInt partitionId;
+        final OptionalLong offset;
+        final Optional<String> includedTopics;
+        final Consumer<byte[], byte[]> consumer;
+        final long timeoutMs;
+        final Time time = Time.SYSTEM;
+
+        Iterator<ConsumerRecord<byte[], byte[]>> recordIter = Collections.emptyIterator();
+
+        public ConsumerWrapper(Optional<String> topic,
```

Comment:
I opted to copy the previous logic to make it easier to review and avoid changing any behavior by mistake. I agree this could be refactored but I'd prefer doing that in a follow up PR.